src/shadow.rs

```diff
@@ -0,0 +1,224 @@
+use syntax::ast::*;
+use syntax::codemap::Span;
+use syntax::visit::FnKind;
+
+use rustc::lint::{Context, LintArray, LintPass};
+use utils::{in_external_macro, snippet, span_lint};
+
+declare_lint!(pub SHADOW_SAME, Allow,
+    "rebinding a name to itself, e.g. `let mut x = &mut x`");
+declare_lint!(pub SHADOW_REUSE, Allow,
+    "rebinding a name to an expression that re-uses the original value, e.g. \
+    `let x = x + 1`");
+declare_lint!(pub SHADOW_FOREIGN, Warn,
+    "The name is re-bound without even using the original value");
+
+#[derive(Copy, Clone)]
+pub struct ShadowPass;
+
+impl LintPass for ShadowPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(SHADOW_SAME, SHADOW_REUSE, SHADOW_FOREIGN)
+    }
+
+    fn check_fn(&mut self, cx: &Context, _: FnKind, decl: &FnDecl,
+            block: &Block, _: Span, _: NodeId) {
+        if in_external_macro(cx, block.span) { return; }
+        check_fn(cx, decl, block);
+    }
+}
+
+fn check_fn(cx: &Context, decl: &FnDecl, block: &Block) {
+    let mut bindings = Vec::new();
+    for arg in &decl.inputs {
+        if let PatIdent(_, ident, _) = arg.pat.node {
+            bindings.push(ident.node.name)
+        }
+    }
+    check_block(cx, block, &mut bindings);
+}
+
+fn named(pat: &Pat) -> Option<Name> {
+    if let PatIdent(_, ident, _) = pat.node {
```

Comment:
this doesn't consider more complex patterns like `(a, b)` or `T { x, y }` at the moment...